Monads/Option.cs

```diff
@@ -133,4 +133,138 @@ public class Some<T>(T value) : Option<T>(value) where T : class;
+        public ValueOption<R> Bind<R>(Func<T, R> func) where R : struct
+            => HasValue ? new(func(_value!.Value)) : new();
+
+        /// <summary>
+        /// Performs one of the actions,
+        /// depending on if there exists an underlying value or not.
+        /// </summary>
+        /// <param name="some">action to perform if a value exists</param>
+        /// <param name="none">action to perform if there is no value</param>
+        public void Inspect(Action<T> some, Action none)
+        {
+            if (HasValue)
+                some(_value.Value);
+            else none();
+        }
+
+        /// <summary>
+        /// Performs one of the operations,
+        /// depending on if there exists an underlying value or not.
+        /// </summary>
+        /// <param name="some">operation to perform if the value exists</param>
+        /// <param name="none">operation to perform if there is no value</param>
+        /// <returns>
+        /// If there is a value, it is sent to <paramref name="some"/> and its result is returned.
+        /// If there is no value, the result of <paramref name="none"/> is returned.
+        /// </returns>
+        public R Map<R>(Func<T, R> some, Func<R> none)
+            => HasValue ? some(_value!.Value) : none();
+
+        /// <summary>
+        /// Tries to return the underlying value.
+        /// </summary>
+        /// <param name="orElse">Value to default to if <see langword="this"/> doesn't have an underlying value</param>
+        /// <returns>
+        /// If there is an underlying value, it is returned.
+        /// Otherwise, <paramref name="orElse"/> is returned.
+        /// </returns>
+        public T ValueOrDefault(T orElse) => _value ?? orElse;
+
+        /// <inheritdoc cref="IEquatable{T}.Equals(T)"/>
+        public bool Equals(ValueOption<T>? other)
+            => other is not null && this.GetHashCode() == other.GetHashCode();
+        /// <summary>
+        /// Indicates whether the <see cref="ValueOption{T}"/>s are equal
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if their underlying values are equal or are both <see langword="null"/>.
+        /// <see langword="false"/> otherwise.
+        /// </returns>
+        public static bool operator ==(ValueOption<T>? self, ValueOption<T>? other)
+            => self is null ? other is null : self.Equals(other);
+        /// <summary>
+        /// Indicates whether the <see cref="ValueOption{T}"/>s are not equal
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if their underlying values are not equal or only one is <see langword="null"/>.
+        /// <see langword="false"/> otherwise.
+        /// </returns>
+        public static bool operator !=(ValueOption<T>? self, ValueOption<T>? other) => !(self == other);
+
+        /// <inheritdoc cref="object.GetHashCode"/>
+        public override int GetHashCode() => _value?.GetHashCode() ?? 0;
```

Comment:
Ve need to get the hash code of the underlying value, access through `_value.Value.GetHashCode()` instead
```suggestion
        public override int GetHashCode() => _value?.Value.GetHashCode() ?? 0;
```

Reply:
also nevermind